### PR TITLE
Workaround the infamous coredns feature

### LINKED ISSUE #2239: helm/k8s deployments seem to have problems with DNS resolution;  Mailu expects validated dnssec records
## Description
As the default k8s dns setup is not working correctly with dnssec, we will (like the docker setup) use unbound as resolver, but to allow us to continue using the internal k8s-dns system, we need a dns forwarder setting.

## Replication Steps
Setup mailu with mailu helm chart in k8s

I will provide a suitable fix myself.

### PR DESCRIPTION
**Pull request by nextgens**
## What type of PR?

bug-fix

## What does this PR do?

Ensure that we set the ``DO`` flag on our queries to work around https://github.com/coredns/coredns/issues/5189

Add a FAQ entry to point users in the right direction in other cases (dnsmasq), discourage users from running Mailu without unbound

### Related issue(s)
- closes #2243
- closes #2239
- #2164
- #2163
- #2162
- #2135
- #1988

## Prerequisites
Before we can consider review and merge, please make sure the following list is done and checked.
If an entry in not applicable, you can check it or remove it from the list.

- [x] In case of feature or enhancement: documentation updated accordingly
- [x] Unless it's docs or a minor change: add [changelog](https://mailu.io/master/contributors/workflow.html#changelog) entry file.

### CODE REVIEW

File: core/admin/start.py [hunk 1]
```diff
@@ -37,11 +37,11 @@ def test_DNS():
             try:
                 result = resolver.resolve('example.org', dns.rdatatype.A, dns.rdataclass.IN, lifetime=10)
             except Exception as e:
-                log.critical("Your DNS resolver at %s is not working (%s). Please use another resolver or enable unbound via https://setup.mailu.io.", ns, e);
+                log.critical("Your DNS resolver at %s is not working (%s). Please see https://mailu.io/1.9/faq.html#the-admin-container-won-t-start-and-its-log-says-critical-your-dns-resolver-isn-t-doing-dnssec-validation", ns, e);
```
Comment (Diman0) [inline core/admin/start.py:40]: This should link to master. Master is always available. The 1.9 documentation site will eventually not be available anymore.

File: core/admin/start.py [hunk 2]
```diff
@@ -37,11 +37,11 @@ def test_DNS():
...
             else:
                 if result.response.flags & dns.flags.AD:
                     break
-                log.critical("Your DNS resolver at %s isn't doing DNSSEC validation; Please use another resolver or enable unbound via https://setup.mailu.io.", ns)
+                log.critical("Your DNS resolver at %s isn't doing DNSSEC validation; Please see https://mailu.io/1.9/faq.html#the-admin-container-won-t-start-and-its-log-says-critical-your-dns-resolver-isn-t-doing-dnssec-validation.", ns)
```
Comment (Diman0) [inline core/admin/start.py:44]: This should link to master. Master is always available. The 1.9 documentation site will eventually not be available anymore.

File: setup/static/render.js [hunk 3]
```diff
@@ -83,4 +83,11 @@ $(document).ready(function() {
 			$("#ipv6").hide();
 		}
 	});
+	$("#unbound").change(function() {
```
Comment (Diman0) [inline setup/static/render.js:86]: It should check the property of the input resolver_enabled

File: setup/templates/steps/compose/03_expose.html [hunk 4]
```diff
@@ -40,13 +40,16 @@
   <input class="form-control" type="text" name="subnet6" required value="{{ subnet6 }}:beef::/64">
 </div>
 
-<p>The unbound resolver enables Mailu to do DNSSEC verification, DNS root lookups and caching. This also helps the antispam service not to get blocked by the public or ISP DNS servers.</p>
 <div class="form-check form-check-inline">
+  <p>The unbound resolver enables Mailu to do DNSSEC verification, DNS root lookups and caching. This also helps the antispam service not to get blocked by the public or ISP DNS servers.</p>
   <label class="form-check-label">
     <input class="form-check-input" type="checkbox" name="resolver_enabled" value="true" checked>
```
Comment (Diman0) [inline setup/templates/steps/compose/03_expose.html:46]: Should get a property id="resolver_enabled" which can be used in render.js for showing the warning message element.

**Review by Diman0 — DISMISSED** [review]
See in-line comments. The main functionality works Ok. 
Thank you for adding the additional warning when unbound is unchecked. I hope this will prevent users from unchecking this without realizing the consequences.

I will submit a commit for the requested changes.

**Comment by Diman0** [comment]
@nextgens  For me the PR is fine now. Since I pushed changes, please @mention me if you are happy now with the PR and I will approve it.

File: core/admin/start.py [hunk 5]
```diff
@@ -37,11 +37,11 @@ def test_DNS():
             try:
                 result = resolver.resolve('example.org', dns.rdatatype.A, dns.rdataclass.IN, lifetime=10)
             except Exception as e:
-                log.critical("Your DNS resolver at %s is not working (%s). Please use another resolver or enable unbound via https://setup.mailu.io.", ns, e);
+                log.critical("Your DNS resolver at %s is not working (%s). Please see https://mailu.io/master/faq.html#the-admin-container-won-t-start-and-its-log-says-critical-your-dns-resolver-isn-t-doing-dnssec-validation", ns, e);
```
Comment (ghostwheel42) [inline core/admin/start.py:40]: I think we could (and should) use /version here.
Just change the message
```
log.critical("Your DNS resolver at %s is not working (%s). Please see https://mailu.io/%s/faq.html#the-admin-container-won-t-start-and-its-log-says-critical-your-dns-resolver-isn-t-doing-dnssec-validation", ns, e, version);
```

And load the file on line ~8:
```
try:
    version = '.'.join(open('/version').read().split('.')[:2])
except:
    version = 'master'
```

**Review by ghostwheel42 — CHANGES_REQUESTED** [review]
what about using /version ?

File: core/admin/start.py (hunk 5 above)
Comment (Diman0) [reply core/admin/start.py:40]: As long the user is not on an old version of Mailu (for which there is not a documentation site), this should work.

Comment (nextgens) [reply core/admin/start.py:40]: IMHO that's all stuff that socrates (or its replacement) should do/deal with. Odds are this isn't the only place that would benefit from knowing the version number

**Comment by nextgens** [comment]
@Diman0 I'm happy with it as is, I'd defer the version improvements to when we have socrates' replacement

**Review by Diman0 — APPROVED** [review]
LGTM

**Comment by entrymon** [comment]
Any workaround on how to setup own resolver? 

I reported issue on Core DNS with details but would be good to know if any easy workaround meanwhile.

https://github.com/coredns/coredns/issues/5826

**Comment by nextgens** [comment]
@entrymon you are creating issues everywhere but at the right place. Stop it.

Since you are using helm/k8s, it's on https://github.com/Mailu/helm-charts/ that you should report your problems.

What I gather from the situation is:
 - you are running an old version of coredns that doesn't have the fix (you need at least 1.9.4 to have the fix, you are running 1.9.3)
 - you are presumably running an old version of Mailu without the workaround too. None of the various comments you have left indicate which version of Mailu you are attempting to run

**Comment by entrymon** [comment]
@nextgens  thanks for pointing me to helm but I don't see why it is helm chart issue. 

Also I upgraded my mailu and coreDNS both and was still getting the error so I posted it here.

Here is my MailU container image 
```
Containers:
  admin:
    Container ID:   docker://33984491a2832b6449b3f22bef283dffa15b21d8d15d9a54aeb1ed960a855429
    Image:          mailu/admin:1.9.45
    Image ID:       docker-pullable://mailu/admin@sha256:9b182ebcdde3723c15044c73b28f0d6189913645432ca6365d5b15d8ee398109
```

Here is my coreDNS container image 
```
Containers:
  coredns:
    Container ID:  docker://bfd0f1694c92c90c88fbbdd2cfd7ed0533e4f72aad9bfb05880d0e855cee9f84
    Image:         registry.k8s.io/coredns/coredns:v1.9.4
    Image ID:      docker-pullable://registry.k8s.io/coredns/coredns@sha256:b82e294de6be763f73ae71266c8f5466e7e03c69f3a1de96efd570284d35bb18
```

Appreciate your help.

**Comment by nextgens** [comment]
This is a recurring issue that won't be fixed until the helm charts point to fixed versions and do what we do in Mailu: ship and enable unbound by default.

In your particular case, you seem to have several problems:
- running a buggy version of coredns (and a version of Mailu without workaround): I understand that is has now been fixed
- using an upstream DNS resolver that's either equally broken or just doesn't do DNSSEC validation

You can use the script on #2619 to investigate your upstream DNS resolution and identify where you loose the AD flag.

**Pull request merged**